### PR TITLE
lexicon and masking

### PR DESCRIPTION
**Pull request by Ilya-Krylov**

### CODE REVIEW

File: mmdet/core/evaluation/text_evaluation.py [hunk 1]
```diff
@@ -225,7 +199,7 @@ def parse_gt_objects(gt_annotation, use_transcription):
     return gt_polygons_list, gt_dont_care_polygon_nums, gt_transcriptions
 
 
-def parse_pr_objects(pr_annotation, conf_thr, use_transcription):
+def parse_pr_objects(pr_annotation, conf_thr, use_transcription, word_spotting):
```
Comment (druzhkov-paul) [inline mmdet/core/evaluation/text_evaluation.py]: `word_spotting` is not used inside the function. Why is it here?

File: mmdet/datasets/coco_with_text.py [hunk 2]
```diff
@@ -1,16 +1,40 @@
 import collections
 import copy
 import logging
+import os
 import string
+import subprocess
+import tempfile
 
+import editdistance
```
Comment (druzhkov-paul) [inline mmdet/datasets/coco_with_text.py:9]: Is it an extra 3rd-party dependency? Is it present in requirements.txt?

File: mmdet/datasets/coco_with_text.py [hunk 3]
```diff
@@ -168,43 +267,55 @@ def evaluate(self,
...
+                texts, text_confidences, character_distributions = res[2]
+                per_image_predictions = []
+
+                for bbox, segm, text, text_conf, char_distrib in zip(boxes, segms, texts, text_confidences, character_distributions):
+                    if text:
+                        text = text.upper()
+                        if lexicon:
+                            text, _ = find_in_lexicon(text, lexicon, lexicon_mapping, char_distrib)
+                            # if metric.startswith('word_spotting'):
+                            #     text = strip(text)
```
Comment (druzhkov-paul) [inline mmdet/datasets/coco_with_text.py]: Is it still needed?

File: mmdet/datasets/weighted_editdistance.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    else:
+        c1 = char2num(word1[i])
+        return scores[c1][i]
+
+
+def ed_replace_cost(i, j, word1, word2, scores):
+    c1 = char2num(word1[i])
+    c2 = char2num(word2[j])
+
+    return max(1 - scores[c2][i]/scores[c1][i]*5, 0)
```
Comment (druzhkov-paul) [inline mmdet/datasets/weighted_editdistance.py]:
```suggestion
    return max(1 - scores[c2][i] / scores[c1][i] * 5, 0)
```

File: mmdet/datasets/weighted_editdistance.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+def ed_delete_cost(j, word1, scores):
+    c = char2num(word1[j])
+    return scores[c][j]
+
+
+def ed_insert_cost(i, word1, scores):
+    if i < len(word1) - 1:
+        c1 = char2num(word1[i])
+        c2 = char2num(word1[i + 1])
+        return (scores[c1][i] + scores[c2][i+1])/2
```
Comment (druzhkov-paul) [inline mmdet/datasets/weighted_editdistance.py]:
```suggestion
        return (scores[c1][i] + scores[c2][i+1]) / 2
```

File: mmdet/datasets/weighted_editdistance.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    m = len(word1)
+    n = len(word2)
+    dp = np.zeros((n + 1, m + 1))
+    dp[0, :] = np.arange(m + 1)
+    dp[:, 0] = np.arange(n + 1)
+    for i in range(1, n + 1):
+        for j in range(1, m + 1):
+            delete_cost = ed_delete_cost(j - 1, word1, scores)
+            insert_cost = ed_insert_cost(j - 1, word1, scores)
+            replace_cost = ed_replace_cost(j-1, i-1, word1, word2, scores) if word1[j - 1] != word2[i - 1] else 0
```
Comment (druzhkov-paul) [inline mmdet/datasets/weighted_editdistance.py]:
```suggestion
            replace_cost = ed_replace_cost(j - 1, i - 1, word1, word2, scores) if word1[j - 1] != word2[i - 1] else 0
```

File: mmdet/models/roi_heads/standard_roi_head_with_text.py [hunk 7]
```diff
@@ -108,13 +112,19 @@ def _text_forward(self, x, rois=None, pos_inds=None, bbox_feats=None, matched_gt
             assert bbox_feats is not None
             text_feats = bbox_feats[pos_inds]
 
+        if self.mask_text_features:
+            hard_masks = det_masks > 0.5
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/standard_roi_head_with_text.py:116]: `det_masks` can be `None`.

**Comment by druzhkov-paul** [comment]
run tests

File: mmdet/core/evaluation/text_evaluation.py (hunk 1 above)
Comment (Ilya-Krylov) [reply mmdet/core/evaluation/text_evaluation.py]: fixed

**Comment by Ilya-Krylov** [comment]
run tests

**Review by druzhkov-paul — APPROVED** [review]

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Pull request merged**